job-cleanup/cleanup.py

```diff
@@ -138,46 +147,78 @@ def label_defunct(client, namespace, job):
                               raise_for_status=False)
         if r.status_code != 200:
             # oh well
-            print('Failed to label pod as defunct: '
-                  '{}/{}'.format(namespace, pod.metadata.name),
-                  file=sys.stderr)
+            logger.error(
+                'Failed to label pod as defunct: %s/%s',
+                namespace, pod.metadata.name)
 
 
-def main():
-    client = KubernetesAPIClient()
-    if USE_KUBE_CONFIG:
-        client.load_kube_config()
-    else:
-        client.load_cluster_config()
+def clean_orphaned(client: KubernetesAPIClient, this_app: str):
+    logger.info('Cleaning orphaned job pods')
 
     namespace = get_current_namespace()
-    pod_name = get_current_pod()
 
-    pod = client.get('/api/v1/namespaces/{}/pods/{}', namespace, pod_name)
+    selector = 'app={},job-name'.format(this_app)
+    pods = client.get('/api/v1/namespaces/{}/pods', namespace,
+                      params={'labelSelector': selector})
 
-    app = pod.metadata.labels['app']
-    component = pod.metadata.labels.get('component', None)
-    this_job = pod.metadata.labels.get('job-name', None)
+    orphaned = []
+    for pod in pods['items']:
+        job_name = pod.metadata.labels['job-name']
+        response = client.get('/apis/batch/v1/namespaces/{}/jobs/{}',
+                              namespace, job_name,
+                              raise_for_status=False)
+        if response.status_code != 404:
+            # not orphaned
+            continue
 
-    if pod_is_self and this_job:
-        selector = 'app={},component!={}'.format(app, component)
-    else:
-        selector = 'app={}'.format(app)
+        if pod.status.phase == 'Running':
+            logger.info('Orphaned pod %s is still running and will not be '
+                        'deleted', pod.metadata.name)
+            continue
 
-    jobs = client.get('/apis/batch/v1/namespaces/{}/jobs', namespace,
-                      params={'labelSelector': selector})
+        orphaned.append(pod)
```

Comment:
Why store the whole pod in this list? Couldn't you just have a list of pod names?

Reply:
That would work fine too, I think both are mostly equivalent here.